### PR TITLE
Fixes for sampling penalties 

### PR DESCRIPTION
**Pull request by sraizada-tt**
This PR fixes several issues with sampling penalties in the LLaMA model implementation, including correcting the repetition penalty logic, adding proper type conversions, and improving padding calculations for device operations.

Key Changes:

Fixed repetition penalty threshold from >1 to >0 and corrected the scaling direction (inverse vs normal penalties)
Added bfloat16 type conversions for frequency and presence penalty terms to ensure compatibility with logits
Updated padding calculation to properly handle tile alignment for vocabulary size across multiple devices

### Checklist
galaxy demo tests: https://github.com/tenstorrent/tt-metal/actions/runs/19852602930/job/56894820619

### CODE REVIEW

**Review by tt-rkim — APPROVED** [review]

**Review by djordje-tt — APPROVED** [review]

**Comment by yieldthought** [comment]
This is the first time I've really looked at the sampling code so apologies if this is not the right PR for these comments:
- How does sampling work with traces? [The code](https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/generator.py#L231) seems to key traces based on whether penalties are on or not but will reuse a trace with seed X when called with seed Y. I think we'd have to [modify ttnn.sampling to take the seed as a tensor](https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/tt_sampling.py#L343) if we want this to work correctly as implemented?
- https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/tt_sampling.py#L128 assumes padded_vocab_size is set but the constructor defaults it to None, so this will throw an exception.

**Comment by sraizada-tt** [comment]
> This is the first time I've really looked at the sampling code so apologies if this is not the right PR for these comments:
> 
> * How does sampling work with traces? [The code](https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/generator.py#L231) seems to key traces based on whether penalties are on or not but will reuse a trace with seed X when called with seed Y. I think we'd have to [modify ttnn.sampling to take the seed as a tensor](https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/tt_sampling.py#L343) if we want this to work correctly as implemented?
> * https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/tt_sampling.py#L128 assumes padded_vocab_size is set but the constructor defaults it to None, so this will throw an exception.

1. seed reset is out of trace since it is not run every decode iteration (I need to remove seed from the tt_sampling file)
2. self.padded_vocab_size  is only called for the galaxy models which have a padded_vocab size in args but yes, this assumption needs to be removed

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -635,6 +635,9 @@ def _capture_decode_trace_text(
                 )
             )
             ttnn.end_trace_capture(self.model_args[i].mesh_device, trace_id, cq_id=0)
+
+            if split_enabled:
+                sampling_module.sample(logits=tt_out_trace[i], tt_out_tok=device_inputs[i][0])
```
Comment (yieldthought) [inline models/tt_transformers/tt/generator.py]: If we call this after _capturing_ a trace, aren't we going to count the statistics _again_ when we actually execute the trace? Capturing doesn't actually do anything on device, right? So we'll end up counting token 0 twice?

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: yes, need to zero out stats after this - good catch!

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: or actually -- need to call capture sampling trace instead of capture + execute

**Review by akirby-TT — APPROVED** [review]

**Review by gwangTT — APPROVED** [review]
LGTM

**Review by skhorasganiTT — APPROVED** [review]

**Comment by yieldthought** [comment]
> > This is the first time I've really looked at the sampling code so apologies if this is not the right PR for these comments:
> > 
> > * How does sampling work with traces? [The code](https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/generator.py#L231) seems to key traces based on whether penalties are on or not but will reuse a trace with seed X when called with seed Y. I think we'd have to [modify ttnn.sampling to take the seed as a tensor](https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/tt_sampling.py#L343) if we want this to work correctly as implemented?
> > * https://github.com/tenstorrent/tt-metal/blob/penalties-fixes/models/common/sampling/tt_sampling.py#L128 assumes padded_vocab_size is set but the constructor defaults it to None, so this will throw an exception.
> 
> 1. seed reset is out of trace since it is not run every decode iteration (I need to remove seed from the tt_sampling file)
> 2. self.padded_vocab_size  is only called for the galaxy models which have a padded_vocab size in args but yes, this assumption needs to be removed

As long as it is made clear that you can't pass a different seed in once decode has started this looks good to me.
